cookbook/offscreen-clipboard-write/background.js

```diff
@@ -0,0 +1,56 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The value that will be written to the clipboard.
+const textToCopy = `Hello world!`;
+
+// When the browser action is clicked, `addToClipboard()` will use an offscreen document to write
+// the value of `textToCopy` to the system clipboard.
+chrome.action.onClicked.addListener(async () => {
+  await addToClipboard(textToCopy);
+});
+
+// Solution 1 - As of Jan 2023, service workers cannot directly interact with the system clipboard.
+// To work around this, we'll create an offscreen document and pass it the data we want to write to
+// the clipboard.
+async function addToClipboard(value) {
+  // This pattern show in this if-else ensures that the offscreen document exists before we try to
+  // send it a message. If we didn't await the `hasDocument()` and `createDocument()` calls, the
+  // `sendMessage()` call could be sent before the offscreen document can register it's
+  // `runtime.onMessage` listener.
+  if (await chrome.offscreen.hasDocument()) {
+    console.debug('Offscreen doc already exists.');
+  } else {
+    console.debug('Creating a new offscreen document.');
+
+    await chrome.offscreen.createDocument({
+      url: 'offscreen.html',
+      reasons: [chrome.offscreen.Reason.CLIPBOARD],
+      justification: 'Write text to the clipboard.',
+    });
+  }
+
+  // Now that are sure we have an off document, we can safely dispatch the message.
+  chrome.runtime.sendMessage({
+    type: 'copy-data-to-clipboard',
+    target: 'offscreen-doc',
+    data: value,
+  });
+}
+
+// Solution 2 – Once extension service workers can use the Clipboard API,
+// replace the offscreen document based implementation with something like this.
```

Comment:
We don't have the work for exposing the Clipboard API in extension service workers scheduled but it is (loosely) something we plan to do. 

* [crbug 1160302](https://crbug.com/1160302) tracks a request to expose clipboard functionality in the extension service worker. 
* [crbug 1404835](https://crbug.com/1404835) specifically requests that we relax the Clipboard API's focus requirement for extensions. 

I'm not strongly opposed to removing this comment and (unused) function, but I would like to explain my rationale for including it before doing so. My goal in including it here was to reinforce the idea that developers should plan for the future when leveraging offscreen documents. Invoking an offscreen document requires that the extension provide one or more reasons (enum) and a justification (string) to identify why the extension is using an offscreen document in the first place. This information informs Chrome what kind of lifetime enforcement it should use for the offscreen document, helps provide reviewers with context on why and how an offscreen document is being used, and finally may be used to inform the user when abnormal behavior is detected.

Once a given web platform feature is exposed in extension service workers, developers should migrate the relevant code out of the offscreen document and into the extension service worker. This will allow the extension to perform the same work faster and using less resources. Developers that are not architecting their code with this future migration in mind may end up having to re-write their implementation when this change eventually comes.

Reply:
@jpmedley How does this context on the idea affect your position? 
If you are opposed to this kind of future-proofing of the sample code itself, do you think it might be useful to add some information related to it into the readme, or should we keep the code sample focused on simply showcasing this feature?

Reply:
It's not really resolved, but we're punting until we can have a policy discussion about how we discuss future plans.